File: pkg/templates/typescript/openai-computer-use/lib/agent.ts
import {
  type ResponseItem,
  type ResponseInputItem,
  type ResponseOutputMessage,
  type ResponseFunctionToolCallItem,
  type ResponseFunctionToolCallOutputItem,
  type ResponseComputerToolCall,
  type ResponseComputerToolCallOutputItem,
  type Tool,
} from 'openai/resources/responses/responses';

import * as utils from './utils';
import type { AgentEvent } from './log-events';
import { describeAction, describeBatchActions } from './log-events';
import { batchInstructions, batchComputerTool, computerUseExtraTool } from './toolset';
import type { CuaAction, KernelComputer } from './kernel-computer';

const BATCH_FUNC_NAME = 'batch_computer_actions';
const EXTRA_FUNC_NAME = 'computer_use_extra';
// Keep this shape aligned with CUA and current OpenAI Responses API.
const OPENAI_COMPUTER_TOOL = { type: 'computer' } as unknown as Tool;

export class Agent {
  private model: string;
  private computer: KernelComputer;
  private tools: Tool[];
  private print_steps = true;
  private debug = false;
  private show_images = false;
  private ackCb: (msg: string) => boolean;
  private onEvent: ((event: AgentEvent) => void) | null = null;
  private modelRequestStartedAt: number | null = null;

  constructor(opts: {
    model?: string;
    computer: KernelComputer;
    tools?: Tool[];
    acknowledge_safety_check_callback?: (msg: string) => boolean;
  }) {
    this.model = opts.model ?? 'gpt-5.4';
    this.computer = opts.computer;
    this.ackCb = opts.acknowledge_safety_check_callback ?? ((): boolean => true);

    this.tools = [
      OPENAI_COMPUTER_TOOL,
      batchComputerTool,
      computerUseExtraTool,
      ...(opts.tools ?? []),
    ];
  }

  private debugPrint(...args: unknown[]): void {
    if (this.debug) {
      try {
        console.dir(
          args.map((msg) => utils.sanitizeMessage(msg as ResponseItem)),
          { depth: null },
        );
      } catch {
        console.dir(args, { depth: null });
      }
    }
  }

  private emit(event: AgentEvent['event'], data: Record<string, unknown>): void {
    if (this.print_steps) this.onEvent?.({ event, data });
  }

  private currentModelElapsedMs(): number | null {
    return this.modelRequestStartedAt === null ? null : Date.now() - this.modelRequestStartedAt;
  }

  private extractReasoningText(item: Record<string, unknown>): string {
    const summary = item.summary;
    if (!Array.isArray(summary)) return '';
    const chunks = summary
      .map((part) => {
        if (!part || typeof part !== 'object') return '';
        const text = (part as { text?: unknown }).text;
        return typeof text === 'string' ? text : '';
      })
      .filter(Boolean);
    return chunks.join(' ').trim();
  }

  private extractUserPrompt(item: ResponseInputItem): string | null {
    const message = item as unknown as { role?: unknown; content?: unknown };
    if (message.role !== 'user') return null;
    if (typeof message.content === 'string') return message.content;
    if (!Array.isArray(message.content)) return null;
    const pieces = message.content
      .map((entry) => {
        if (!entry || typeof entry !== 'object') return '';
        const text = (entry as { text?: unknown }).text;
        return typeof text === 'string' ? text : '';
      })
      .filter(Boolean);
    return pieces.length > 0 ? pieces.join(' ') : null;
  }

  private async handleItem(item: ResponseItem): Promise<ResponseItem[]> {
    const itemType = (item as { type?: string }).type;
    if (itemType === 'reasoning') {
      const text = this.extractReasoningText(item as unknown as Record<string, unknown>);
      if (text) this.emit('reasoning_delta', { text });
    }

    if (item.type === 'message') {
      const msg = item as ResponseOutputMessage;
      const c = msg.content;
      if (msg.role === 'assistant' && Array.isArray(c)) {
        for (const part of c) {
          if (part && typeof part === 'object' && 'text' in part && typeof part.text === 'string') {
            this.emit('text_delta', { text: part.text });
          }
        }
        this.emit('text_done', {});
      }
    }

    if (item.type === 'function_call') {
      const fc = item as ResponseFunctionToolCallItem;
      const argsObj = JSON.parse(fc.arguments) as Record<string, unknown>;
      if (fc.name === BATCH_FUNC_NAME && Array.isArray(argsObj.actions)) {
        const actions = argsObj.actions.filter(
          (action): action is Record<string, unknown> =>
            typeof action === 'object' && action !== null,
        );
        const elapsedMs = this.currentModelElapsedMs();
        this.emit('action', {
          action_type: 'batch',
          description: describeBatchActions(actions),
          action: { type: 'batch', actions },
          ...(elapsedMs === null ? {} : { elapsed_ms: elapsedMs }),
        });
      } else {
        const elapsedMs = this.currentModelElapsedMs();
        this.emit('action', {
          action_type: fc.name,
          description: `${fc.name}(${JSON.stringify(argsObj)})`,
          action: argsObj,
          ...(elapsedMs === null ? {} : { elapsed_ms: elapsedMs }),
        });
      }

      if (fc.name === BATCH_FUNC_NAME) {
        return this.handleBatchCall(fc.call_id, argsObj);
      }
      if (fc.name === EXTRA_FUNC_NAME) {
        return this.handleExtraCall(fc.call_id, argsObj);
      }

      return [
        {
          type: 'function_call_output',
          call_id: fc.call_id,
          output: `Unsupported function call: ${fc.name}`,
        } as unknown as ResponseFunctionToolCallOutputItem,
      ];
    }

    if (item.type === 'computer_call') {
      const cc = item as ResponseComputerToolCall & {
        action?: Record<string, unknown>;
        actions?: Array<Record<string, unknown>>;
      };
      const actionList = Array.isArray(cc.actions)
        ? cc.actions
        : cc.action
          ? [cc.action]
          : [];

      const elapsedMs = this.currentModelElapsedMs();
      const actionType =
        actionList.length === 1 ? String(actionList[0]?.type ?? 'unknown') : 'batch';
      const description =
        actionList.length === 1
          ? describeAction(actionType, actionList[0] ?? {})
          : describeBatchActions(actionList);
      const actionPayload =
        actionList.length === 1 ? (actionList[0] ?? {}) : { type: 'batch', actions: actionList };
      this.emit('action', {
        action_type: actionType,
        description,
        action: actionPayload,
        ...(elapsedMs === null ? {} : { elapsed_ms: elapsedMs }),
      });
      await this.computer.batchActions(actionList as CuaAction[]);

      const screenshot = await this.computer.screenshot();
      this.emit('screenshot', { captured: true, bytes_base64: screenshot.length });

      const pending = cc.pending_safety_checks ?? [];
      for (const check of pending) {
        const msg = check.message ?? '';
        if (!this.ackCb(msg)) throw new Error(`Safety check failed: ${msg}`);
      }

      const currentUrl = await this.computer.getCurrentUrl();
      utils.checkBlocklistedUrl(currentUrl);

      const screenshotOutput = {
        type: 'computer_screenshot',
        image_url: `data:image/png;base64,${screenshot}`,
      } as unknown as ResponseComputerToolCallOutputItem['output'];
      (screenshotOutput as { current_url?: string }).current_url = currentUrl;

      const out: Omit<ResponseComputerToolCallOutputItem, 'id'> = {
        type: 'computer_call_output',
        call_id: cc.call_id,
        acknowledged_safety_checks: pending,
        output: screenshotOutput,
      };
      return [out as ResponseItem];
    }

    return [];
  }

  private async handleBatchCall(
    callId: string,
    argsObj: Record<string, unknown>,
  ): Promise<ResponseItem[]> {
    const actions = Array.isArray(argsObj.actions) ? (argsObj.actions as CuaAction[]) : [];
    await this.computer.batchActions(actions);

    let statusText = 'Actions executed successfully.';
    const terminalReadAction = this.batchTerminalReadAction(actions);
    if (terminalReadAction === 'url') {
      try {
        const currentUrl = await this.computer.getCurrentUrl();
        statusText = `Actions executed successfully. Current URL: ${currentUrl}`;
      } catch (error) {
        statusText = `Actions executed, but url() failed: ${error instanceof Error ? error.message : String(error)}`;
      }
    }

    const outputItems: Array<Record<string, unknown>> = [{ type: 'text', text: statusText }];
    if (terminalReadAction !== 'url') {
      const screenshot = await this.computer.screenshot();
      outputItems.push({
        type: 'image_url',
        image_url: `data:image/png;base64,${screenshot}`,
        detail: 'original',
      });
    }
    return [
      {
        type: 'function_call_output',
        call_id: callId,
        output: JSON.stringify(outputItems),
      } as unknown as ResponseFunctionToolCallOutputItem,
    ];
  }

  private async handleExtraCall(
    callId: string,
    argsObj: Record<string, unknown>,
  ): Promise<ResponseItem[]> {
    const action = typeof argsObj.action === 'string' ? argsObj.action : '';
    const url = typeof argsObj.url === 'string' ? argsObj.url : '';
    let statusText = '';
    if (action === 'goto') {
      await this.computer.batchActions([{ type: 'goto', url }]);
      statusText = 'goto executed successfully.';
    } else if (action === 'back') {
      await this.computer.batchActions([{ type: 'back' }]);
      statusText = 'back executed successfully.';
    } else if (action === 'url') {
      const currentUrl = await this.computer.getCurrentUrl();
      statusText = `Current URL: ${currentUrl}`;
    } else {
      statusText = `unknown ${EXTRA_FUNC_NAME} action: ${action}`;
    }

    const outputItems: Array<Record<string, unknown>> = [{ type: 'text', text: statusText }];
    if (action !== 'url') {
      const screenshot = await this.computer.screenshot();
      outputItems.push({
        type: 'image_url',
        image_url: `data:image/png;base64,${screenshot}`,
        detail: 'original',
      });
    }
    return [
      {
        type: 'function_call_output',
        call_id: callId,
        output: JSON.stringify(outputItems),
      } as unknown as ResponseFunctionToolCallOutputItem,
    ];
  }

  private batchTerminalReadAction(actions: CuaAction[]): '' | 'url' | 'screenshot' {
    if (actions.length === 0) return '';
    const lastType = actions[actions.length - 1]?.type;
    if (lastType === 'url' || lastType === 'screenshot') return lastType;
    return '';
  }

  async runFullTurn(opts: {
    messages: ResponseInputItem[];
    print_steps?: boolean;
    debug?: boolean;
    show_images?: boolean;
    onEvent?: (event: AgentEvent) => void;
  }): Promise<ResponseItem[]> {
    this.print_steps = opts.print_steps ?? true;
    this.debug = opts.debug ?? false;
    this.show_images = opts.show_images ?? false;
    this.onEvent = opts.onEvent ?? null;
    const newItems: ResponseItem[] = [];
    let turns = 0;

    for (const message of opts.messages) {
      const prompt = this.extractUserPrompt(message);
      if (prompt) this.emit('prompt', { text: prompt });
    }

    try {
      while (
        newItems.length === 0 ||
        (newItems[newItems.length - 1] as ResponseItem & { role?: string }).role !== 'assistant'
      ) {
        turns += 1;
        const inputMessages = [...opts.messages];

        this.debugPrint(...inputMessages, ...newItems);
        this.modelRequestStartedAt = Date.now();
        const response = await utils.createResponse({
          model: this.model,
          input: [...inputMessages, ...newItems],
          tools: this.tools,
          truncation: 'auto',
          reasoning: {
            effort: 'low',
            summary: 'concise',
          },
          instructions: batchInstructions,
        });
        if (!response.output) throw new Error('No output from model');
        for (const msg of response.output as ResponseItem[]) {
          newItems.push(msg, ...(await this.handleItem(msg)));
        }
        this.modelRequestStartedAt = null;
        this.emit('turn_done', { turn: turns });
      }
    } catch (error) {
      this.modelRequestStartedAt = null;
      this.emit('error', { message: error instanceof Error ? error.message : String(error) });
      throw error;
    }
    this.emit('run_complete', { turns });

    return !this.show_images
      ? newItems.map((msg) => utils.sanitizeMessage(msg) as ResponseItem)
      : newItems;
  }
}

File: pkg/templates/python/openai-computer-use/agent/agent.py
import json
import time
from typing import Any, Callable
from computers.kernel_computer import (
    KernelComputer,
    _describe_action,
    _describe_batch_actions,
)
from utils import (
    create_response,
    show_image,
    pp,
    sanitize_message,
    check_blocklisted_url,
)

BATCH_FUNC_NAME = "batch_computer_actions"
EXTRA_FUNC_NAME = "computer_use_extra"

BATCH_INSTRUCTIONS = """You have three ways to perform actions:
1. The standard computer tool — use for single actions when you need screenshot feedback after each step.
2. batch_computer_actions — use to execute multiple actions at once when you can predict the outcome.
3. computer_use_extra — use high-level browser actions: goto, back, and url.

ALWAYS prefer batch_computer_actions when performing predictable sequences like:
- Clicking a text field, typing text, and pressing Enter
- Any sequence where you don't need to see intermediate results

Use computer_use_extra for:
- action="goto" only when changing the page URL
- action="back" to go back in history
- action="url" to read the exact current URL

When interacting with page content (search boxes, forms, chat inputs):
- Click the target input first, then type.
- Do not use URL-navigation actions for in-page text entry."""

BATCH_TOOL = {
    "type": "function",
    "name": BATCH_FUNC_NAME,
    "description": (
        "Execute multiple computer actions in sequence without waiting for "
        "screenshots between them. Use this when you can predict the outcome of a "
        "sequence of actions without needing intermediate visual feedback. After all "
        "actions execute, a single screenshot is taken and returned.\n\n"
        "PREFER this over individual computer actions when:\n"
        "- Typing text followed by pressing Enter\n"
        "- Clicking a field and then typing into it\n"
        "- Any sequence where intermediate screenshots aren't needed\n\n"
        "Constraint: return-value actions (url, screenshot) can appear at most once "
        "and only as the final action in the batch."
    ),
    "parameters": {
        "type": "object",
        "properties": {
            "actions": {
                "type": "array",
                "description": "Ordered list of actions to execute",
                "items": {
                    "type": "object",
                    "properties": {
                        "type": {
                            "type": "string",
                            "enum": [
                                "click",
                                "double_click",
                                "type",
                                "keypress",
                                "scroll",
                                "move",
                                "drag",
                                "wait",
                                "goto",
                                "back",
                                "url",
                                "screenshot",
                            ],
                        },
                        "x": {"type": "number"},
                        "y": {"type": "number"},
                        "text": {"type": "string"},
                        "url": {"type": "string"},
                        "keys": {"type": "array", "items": {"type": "string"}},
                        "hold_keys": {"type": "array", "items": {"type": "string"}},
                        "button": {"type": "string"},
                        "scroll_x": {"type": "number"},
                        "scroll_y": {"type": "number"},
                    },
                    "required": ["type"],
                },
            },
        },
        "required": ["actions"],
    },
    "strict": False,
}

EXTRA_TOOL = {
    "type": "function",
    "name": EXTRA_FUNC_NAME,
    "description": "High-level browser actions for navigation and URL retrieval.",
    "parameters": {
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": ["goto", "back", "url"],
                "description": "Action to perform: goto, back, or url.",
            },
            "url": {
                "type": "string",
                "description": "Required when action is goto. Fully qualified URL to navigate to.",
            },
        },
        "required": ["action"],
    },
    "strict": False,
}

# Keep this shape aligned with CUA and current OpenAI Responses API.
OPENAI_COMPUTER_TOOL = {"type": "computer"}


class Agent:
    """An agent that uses OpenAI CUA with Kernel's native computer control API."""

    def __init__(
        self,
        model="gpt-5.4",
        computer: KernelComputer = None,
        tools: list[dict] = [],
        acknowledge_safety_check_callback: Callable = lambda message: False,
    ):
        self.model = model
        self.computer = computer
        self.tools = list(tools)
        self.print_steps = True
        self.debug = False
        self.show_images = False
        self.on_event: Callable[[dict], None] | None = None
        self._model_request_started_at: float | None = None
        self.acknowledge_safety_check_callback = acknowledge_safety_check_callback

        if computer:
            self.tools += [
                dict(OPENAI_COMPUTER_TOOL),
                BATCH_TOOL,
                EXTRA_TOOL,
            ]

    def debug_print(self, *args):
        if self.debug:
            pp(*args)

    def _emit_event(self, event: str, data: dict | None = None) -> None:
        if self.print_steps and self.on_event:
            self.on_event({"event": event, "data": data or {}})

    def _current_model_elapsed_ms(self) -> int | None:
        if self._model_request_started_at is None:
            return None
        return int((time.time() - self._model_request_started_at) * 1000)

    def _extract_reasoning_text(self, item: dict[str, Any]) -> str:
        summary = item.get("summary")
        if not isinstance(summary, list):
            return ""
        pieces: list[str] = []
        for part in summary:
            if not isinstance(part, dict):
                continue
            text = part.get("text")
            if isinstance(text, str) and text:
                pieces.append(text)
        return " ".join(pieces).strip()

    def _extract_prompt_text(self, item: dict[str, Any]) -> str | None:
        if item.get("role") != "user":
            return None
        content = item.get("content")
        if isinstance(content, str):
            return content
        if not isinstance(content, list):
            return None
        parts: list[str] = []
        for entry in content:
            if not isinstance(entry, dict):
                continue
            text = entry.get("text")
            if isinstance(text, str) and text:
                parts.append(text)
        return " ".join(parts) if parts else None

    def _batch_terminal_read_action(self, actions: list[dict[str, Any]]) -> str:
        if not actions:
            return ""
        action_type = str(actions[-1].get("type", ""))
        if action_type in ("url", "screenshot"):
            return action_type
        return ""

    def handle_item(self, item):
        """Handle each item; may cause a computer action + screenshot."""
        if item["type"] == "reasoning":
            text = self._extract_reasoning_text(item)
            if text:
                self._emit_event("reasoning_delta", {"text": text})

        if item["type"] == "message":
            if item.get("role") == "assistant":
                content = item.get("content", [])
                if isinstance(content, list):
                    for part in content:
                        if isinstance(part, dict) and isinstance(part.get("text"), str):
                            self._emit_event("text_delta", {"text": part["text"]})
                self._emit_event("text_done", {})

        if item["type"] == "function_call":
            name, args = item["name"], json.loads(item["arguments"])
            elapsed_ms = self._current_model_elapsed_ms()
            if name == BATCH_FUNC_NAME:
                actions = args.get("actions", [])
                if isinstance(actions, list):
                    typed_actions = [a for a in actions if isinstance(a, dict)]
                    payload = {
                        "action_type": "batch",
                        "description": _describe_batch_actions(typed_actions),
                        "action": {"type": "batch", "actions": typed_actions},
                    }
                    if elapsed_ms is not None:
                        payload["elapsed_ms"] = elapsed_ms
                    self._emit_event(
                        "action",
                        payload,
                    )
            else:
                payload = {
                    "action_type": name,
                    "description": f"{name}({json.dumps(args)})",
                    "action": args,
                }
                if elapsed_ms is not None:
                    payload["elapsed_ms"] = elapsed_ms
                self._emit_event(
                    "action",
                    payload,
                )

            if name == BATCH_FUNC_NAME:
                return self._handle_batch_call(item["call_id"], args)
            if name == EXTRA_FUNC_NAME:
                return self._handle_extra_call(item["call_id"], args)

            return [
                {
                    "type": "function_call_output",
                    "call_id": item["call_id"],
                    "output": f"Unsupported function call: {name}",
                }
            ]

        if item["type"] == "computer_call":
            elapsed_ms = self._current_model_elapsed_ms()
            actions = item.get("actions")
            if not isinstance(actions, list):
                single = item.get("action")
                actions = [single] if isinstance(single, dict) else []
            typed_actions = [a for a in actions if isinstance(a, dict)]

            if len(typed_actions) == 1:
                action_type = str(typed_actions[0].get("type", "unknown"))
                action_payload: dict[str, Any] = typed_actions[0]
                description = _describe_action(
                    action_type,
                    {k: v for k, v in typed_actions[0].items() if k != "type"},
                )
            else:
                action_type = "batch"
                action_payload = {"type": "batch", "actions": typed_actions}
                description = _describe_batch_actions(typed_actions)

            payload = {
                "action_type": action_type,
                "description": description,
                "action": action_payload,
            }
            if elapsed_ms is not None:
                payload["elapsed_ms"] = elapsed_ms
            self._emit_event("action", payload)
            self.computer.batch_actions(typed_actions)

            screenshot_base64 = self.computer.screenshot()
            self._emit_event(
                "screenshot",
                {"captured": True, "bytes_base64": len(screenshot_base64)},
            )
            if self.show_images:
                show_image(screenshot_base64)

            pending_checks = item.get("pending_safety_checks", [])
            for check in pending_checks:
                message = check["message"]
                if not self.acknowledge_safety_check_callback(message):
                    raise ValueError(
                        f"Safety check failed: {message}. Cannot continue with unacknowledged safety checks."
                    )

            call_output = {
                "type": "computer_call_output",
                "call_id": item["call_id"],
                "acknowledged_safety_checks": pending_checks,
                "output": {
                    "type": "computer_screenshot",
                    "image_url": f"data:image/png;base64,{screenshot_base64}",
                },
            }

            if self.computer.get_environment() == "browser":
                current_url = self.computer.get_current_url()
                check_blocklisted_url(current_url)
                call_output["output"]["current_url"] = current_url

            return [call_output]
        return []

    def _handle_batch_call(self, call_id, args):
        actions = args.get("actions", [])
        if not isinstance(actions, list):
            actions = []
        self.computer.batch_actions(actions)
        status_text = "Actions executed successfully."
        terminal_action = self._batch_terminal_read_action(actions if isinstance(actions, list) else [])
        if terminal_action == "url":
            try:
                current_url = self.computer.get_current_url()
                status_text = f"Actions executed successfully. Current URL: {current_url}"
            except Exception as exc:
                status_text = f"Actions executed, but url() failed: {exc}"
        output_items: list[dict[str, Any]] = [{"type": "text", "text": status_text}]
        if terminal_action != "url":
            screenshot_base64 = self.computer.screenshot()
            output_items.append(
                {
                    "type": "image_url",
                    "image_url": f"data:image/png;base64,{screenshot_base64}",
                    "detail": "original",
                }
            )
        return [
            {
                "type": "function_call_output",
                "call_id": call_id,
                "output": json.dumps(output_items),
            }
        ]

    def _handle_extra_call(self, call_id, args):
        action = args.get("action", "")
        url = args.get("url", "")
        if action == "goto":
            self.computer.batch_actions([{"type": "goto", "url": url}])
            status_text = "goto executed successfully."
        elif action == "back":
            self.computer.batch_actions([{"type": "back"}])
            status_text = "back executed successfully."
        elif action == "url":
            status_text = f"Current URL: {self.computer.get_current_url()}"
        else:
            status_text = f"unknown {EXTRA_FUNC_NAME} action: {action}"

        output_items: list[dict[str, Any]] = [{"type": "text", "text": status_text}]
        if action != "url":
            screenshot_base64 = self.computer.screenshot()
            output_items.append(
                {
                    "type": "image_url",
                    "image_url": f"data:image/png;base64,{screenshot_base64}",
                    "detail": "original",
                }
            )
        return [
            {
                "type": "function_call_output",
                "call_id": call_id,
                "output": json.dumps(output_items),
            }
        ]

    def run_full_turn(
        self,
        input_items,
        print_steps=True,
        debug=False,
        show_images=False,
        on_event: Callable[[dict], None] | None = None,
    ):
        self.print_steps = print_steps
        self.debug = debug
        self.show_images = show_images
        self.on_event = on_event
        new_items = []
        turns = 0

        for message in input_items:
            if isinstance(message, dict):
                prompt = self._extract_prompt_text(message)
                if prompt:
                    self._emit_event("prompt", {"text": prompt})

        try:
            while new_items[-1].get("role") != "assistant" if new_items else True:
                turns += 1
                self.debug_print([sanitize_message(msg) for msg in input_items + new_items])

                self._model_request_started_at = time.time()
                response = create_response(
                    model=self.model,
                    input=input_items + new_items,
                    tools=self.tools,
                    truncation="auto",
                    reasoning={
                        "effort": "low",
                        "summary": "concise",
                    },
                    instructions=BATCH_INSTRUCTIONS,
                )
                self.debug_print(response)

                if "output" not in response:
                    if self.debug:
                        print(response)
                    raise ValueError("No output from model")

                new_items += response["output"]
                for item in response["output"]:
                    new_items += self.handle_item(item)
                self._model_request_started_at = None
                self._emit_event("turn_done", {"turn": turns})
        except Exception as exc:
            self._model_request_started_at = None
            self._emit_event("error", {"message": str(exc)})
            raise

        self._emit_event("run_complete", {"turns": turns})
        return new_items
